SOW-09 - Low-Level Component Documentation implementation (#550)
midnightntwrk/midnight-ledger: spec/zkir.md

| `num_inputs` | `u32` | Number of initial memory elements. Memory indices `0 .. num_inputs` are the circuit inputs, taken from `ProofPreimage::inputs`. |
| `do_communications_commitment` | `bool` | Whether the circuit binds a commitment over its inputs and outputs ([§1.5](#15-communications-commitment)). |
| `instructions` | `Arc<Vec<Instruction>>` | The instruction list, executed in order. |

### 1.2 The memory tape

* Memory is a `Vec` of **native field elements** (`Fr`).
* It starts as a copy of `ProofPreimage::inputs` (length `num_inputs`).
* Instructions execute in order. Each instruction reads operands by index and
  appends its output(s) to the end of memory. Outputs therefore occupy the
  next consecutive indices.
* An operand of type **`Index`** is just a `u32` position into this tape.
  Reading an out-of-range index is a runtime error (`index out of bounds`).

Because outputs are appended, an instruction can only reference values produced
**before** it. The number of values each instruction appends is fixed and
listed per-instruction below (most push 1; EC ops and `DivModPowerOfTwo` /

> `PersistentHash` push 2; pure constraints / assertions push 0).

Reviewer: ```suggestion
listed per-instruction below.
```

### 1.3 The field and the embedded curve

* **Native field (`Fr`)** — the **BLS12-381 scalar field** (`outer::Scalar` in
  `transient-crypto`). Every memory element is a native field element.

>   `FR_BITS = 255`, `FR_BYTES = 32`, `FR_BYTES_STORED = 31`.

Reviewer: ```suggestion
```

* **Embedded curve** — **Jubjub** (`EmbeddedGroupAffine`; the Jubjub base field
  is the BLS12-381 scalar field, which is why curve coordinates fit directly
  in native memory elements). In v2, a curve point is represented as **two**

>   memory elements: its affine `x` and `y` coordinates.

Reviewer: ```suggestion
* **Embedded curve** — ZKIRv2 supports operations over the embedded curve, **Jubjub**. Jubjub is the native curve, whose base field is compact's 'native field'. This enables efficient operations over this curve. A curve point is represented as **two**
  memory elements: its affine `x` and `y` coordinates.
```

* **Embedded scalar field** — the Jubjub scalar field (`EmbeddedFr`), used as

>   the scalar argument to `EcMul` / `EcMulGenerator`.

Reviewer: ```suggestion
* **Embedded scalar field** — the Jubjub scalar field (`EmbeddedFr`), used as 
  the scalar argument to `EcMul` / `EcMulGenerator`. Scalars are carried in 
  native `Fr` memory slots, and any value exceeding the Jubjub scalar-field 
  order is reduced modulo it.
```

### 1.4 Public inputs, transcripts, and guards

A ZKIR proof is verified against a list of **public inputs (PIs)**. The PI
vector is assembled during preprocessing in this order:

1. `preimage.binding_input` (always first).
2. The communications-commitment value, **iff**
   `do_communications_commitment`.
3. One entry per `DeclarePubInput` instruction, in execution order.

> 

Reviewer: ```suggestion

The `DeclarePubInput` entries are how the circuit binds itself to a specific sequence of **Impact VM operations**: each operation's arguments are pinned into the PI vector, and the proof attests that the circuit produced exactly that sequence. 

  > Intuitively, the public inputs *are* the Impact VM operations the proof
  > commits to: a valid proof attests that some ZKIR execution produced
  > exactly that sequence, and the Impact VM then executes those operations
  > on chain. The proof therefore gates *which* instructions reach the VM -
  > they are run only if the proof verifies.
  
```

the instruction takes its "inactive" path: `PublicInput` / `PrivateInput` push
`0` without advancing the transcript cursor, and `PiSkip` records a skip.
This is how conditional (guarded) Impact operations are encoded.

### 1.5 Communications commitment

When `do_communications_commitment` is set, the circuit binds a
Pedersen-style `transientCommit` over the concatenation `inputs ‖ outputs`
(where `outputs` are the values collected by `Output` instructions).
Preprocessing recomputes this commitment from the preimage's
`communications_commitment` `(value, randomness)` and fails with
`Communications commitment mismatch` if it disagrees. The commitment value is
added as a public input (position 2 above).

### 1.6 Minor versions (V0 vs V1)

`IrMinorVersion` is `V0` or `V1` (default `V1`). The two differ only in how a
couple of instructions enforce bit bounds in-circuit. For example, `V0`
`CondSelect` derives the selector via `is_zero` (negating the bit to force the
bound), and `V0` `ConstrainBits` lowers slightly differently from `V1`.
Functional results are equivalent; `V1` is the optimised lowering. New
circuits use `V1`.

> ## 2. Operand and value conventions

Reviewer: Probably this does not deserve its own section. I'd call it 3.1

| Convention | Meaning |
|---|---|
| `Index` | A `u32` memory position. The value stored there is a native field element. |

> | boolean operand | An `Index` whose value **must** be `0` or `1`. Reading a non-boolean where a boolean is required is a runtime error (`Expected boolean`); in-circuit, callers are responsible for the `0/1` precondition (see "UB" notes). |

Reviewer: The use of boolean always enforces that the value is '0/1' in-circuit.

Author: I've simplified the text

| point operand | A **pair** of indices `(x, y)`. The pair must be a valid Jubjub point or preprocessing errors (`point not on curve`). |

> | scalar operand | An `Index` interpreted as an embedded-curve (Jubjub) scalar. |

Reviewer: ```suggestion
| scalar operand | An `Index` interpreted as a Jubjub scalar. |
```

| `imm` | A native field constant embedded in the instruction (`LoadImm`). |
| `alignment` | A field-aligned-binary `Alignment` describing how inputs pack into bytes (`PersistentHash`). |

**Serialised form.**  Instructions serialise (serde) as a tagged object with
key `"op"` and snake-cased variant name, e.g. `Add { a, b }` →
`{"op": "add", "a": 3, "b": 7}`.  Examples below use this form.

## 3. Instruction reference (IR v2)

There are **26** v2 instructions.  Each entry lists operands, the number of
values pushed to memory ("outputs"), semantics, and preconditions / undefined
behaviour (UB).  "UB" means the instruction does not itself constrain the
precondition — the caller must guarantee it, or the proof / preprocessing may
be unsound or error.

### 3.1 Field arithmetic

#### `Add { a, b }`
* **Operands:** `a: Index`, `b: Index`. **Outputs:** 1 — `a + b`.
* **Semantics:** addition in the native prime field.

#### `Mul { a, b }`
* **Operands:** `a: Index`, `b: Index`. **Outputs:** 1 — `a * b`.
* **Semantics:** multiplication in the native prime field.

#### `Neg { a }`
* **Operands:** `a: Index`. **Outputs:** 1 — `-a`.

> * **Semantics:** additive negation in the native prime field.

Reviewer 2: ```suggestion
* **Semantics:** negation in the native prime field.
```

### 3.2 Boolean and selection

#### `Not { a }`
* **Operands:** `a: Index` (boolean). **Outputs:** 1 — `!a`.
* **Precondition:** `a ∈ {0,1}` (runtime-checked as boolean during preprocessing).

#### `TestEq { a, b }`
* **Operands:** `a: Index`, `b: Index`. **Outputs:** 1 — boolean `a == b`.
* **Semantics:** returns `1` if equal, else `0`. (Compare with `ConstrainEq`,
  which *asserts* equality rather than computing a boolean.)

#### `CondSelect { bit, a, b }`
* **Operands:** `bit: Index` (boolean), `a: Index`, `b: Index`. **Outputs:** 1.
* **Semantics:** pushes `a` if `bit == 1`, else `b`.

> * **UB:** `bit` must be `0` or `1`. (Lowering differs between `V0` / `V1`;

Reviewer: This is a precondition

  see §1.6.)

### 3.3 Constraints and assertions (push **no** outputs)

#### `Assert { cond }`
* **Operands:** `cond: Index` (boolean). **Outputs:** none.
* **Semantics:** requires `cond == 1`. Fails preprocessing with
  `Failed direct assertion` otherwise; in-circuit, asserts non-zero.

> * **UB:** `cond` must be `0` or `1`.

Reviewer: precondition

#### `ConstrainEq { a, b }`
* **Operands:** `a: Index`, `b: Index`. **Outputs:** none.
* **Semantics:** constrains `a == b`. Fails with `Failed equality constraint`
  otherwise.

#### `ConstrainToBoolean { var }`
* **Operands:** `var: Index`. **Outputs:** none.

> * **Semantics:** constrains `var ∈ {0,1}`.

Reviewer: We are missing some consistency on the semantics field. Sometimes it says how it fails, others it doesn't. I'd suggest that we add another field for the error if we want to expose that in the docs.

#### `ConstrainBits { var, bits }`
* **Operands:** `var: Index`, `bits: u32`. **Outputs:** none.
* **Semantics:** constrains `var` to fit in `bits` bits (i.e. `var < 2^bits`).

> * **Precondition:** `bits < FR_BITS` (255). A larger bound errors with

Reviewer: ```suggestion
* **Precondition:** `|bits| < FR_BITS` (255). A larger bound errors with
```

  `Excessive bit bound`.

### 3.4 Bit decomposition and reconstruction

#### `DivModPowerOfTwo { var, bits }`
* **Operands:** `var: Index`, `bits: u32`. **Outputs:** 2 — pushed in this
  order:
  1. `var >> bits` (the quotient / high bits)
  2. `var & ((1 << bits) - 1)` (the remainder / low `bits` bits)

> * **Precondition:** `bits ≤ FR_BYTES_STORED * 8` (248). Larger errors with

Reviewer: ```suggestion
* **Precondition:** `|bits| ≤ FR_BYTES_STORED * 8` (248). Larger errors with
```

  `Excessive bit count`.

#### `ReconstituteField { divisor, modulus, bits }`
* **Operands:** `divisor: Index`, `modulus: Index`, `bits: u32`. **Outputs:**
  1 — `(divisor << bits) | modulus`.
* **Semantics:** inverse of `DivModPowerOfTwo`. Guarantees `modulus < 2^bits`
  and that the reconstituted value does not overflow the field (errors:
  `Reconstituted element overflows field`, or `Excessive bit count` if
  `bits > 248`).
* **Example (round-trip):**

  ```jsonc
  // split memory[5] into high/low at 8 bits, then rebuild it
  {"op": "div_mod_power_of_two", "var": 5, "bits": 8}      // pushes hi@N, lo@N+1
  {"op": "reconstitute_field", "divisor": N, "modulus": N+1, "bits": 8}  // == memory[5]
  ```

#### `LessThan { a, b, bits }`
* **Operands:** `a: Index`, `b: Index`, `bits: u32`. **Outputs:** 1 — boolean
  `a < b`.
* **Semantics:** compares `a` and `b` as `bits`-bit unsigned integers.

> * **UB:** `a` and `b` must each fit in `bits` bits; otherwise the comparison

Reviewer: ```suggestion
* **Precondition:** `a` and `b` must each fit in `bits` bits; otherwise the comparison
```

* **Semantics:** **Poseidon** hash (over the native scalar field) of the
  listed memory elements. Circuit-friendly and cheap in-circuit; use for
  transient / ephemeral hashing.

#### `PersistentHash { alignment, inputs }`
* **Operands:** `alignment: Alignment`, `inputs: Vec<Index>`. **Outputs:** 2 —
  the hash digest in the binary (field-aligned) format.
* **Semantics:** **SHA-256** hash. Inputs are first parsed according to
  `alignment` into a field-aligned-binary (FAB) byte representation
  ([field-aligned-binary.md](./field-aligned-binary.md)), then SHA-256 is
  taken over those bytes; the 32-byte digest is returned as 2 field elements.
  Errors with `Inputs did not match alignment` if `inputs` don't conform to
  `alignment`.
* **Note:** much more expensive in-circuit than `TransientHash` (SHA-256 vs
  Poseidon); use where the digest must be a standard, stable hash across
  contexts (e.g. on-chain identifiers, interop).

### 3.7 Elliptic-curve operations (Jubjub)

Curve points are `(x, y)` index pairs; each of these pushes **2** outputs
(`c_x`, `c_y`).

#### `EcAdd { a_x, a_y, b_x, b_y }`
* **Outputs:** 2 — point `a + b`.

> * **UB:** `(a_x, a_y)` and `(b_x, b_y)` must each be valid Jubjub points

Reviewer: Prerequisite

Author: Updated.

  (else `point not on curve`).

#### `EcMul { a_x, a_y, scalar }`
* **Operands:** point `(a_x, a_y)`, `scalar: Index`. **Outputs:** 2 —
  `scalar · a`.

> * **UB:** `(a_x, a_y)` must be a valid curve point.

Reviewer: Prerequisite. This is not undefined behaviour.

Author: Do you mean this will always be a valid curve point so not worth mentioning?

#### `EcMulGenerator { scalar }`
* **Operands:** `scalar: Index`. **Outputs:** 2 — `scalar · G` (the fixed
  Jubjub generator).

#### `HashToCurve { inputs }`
* **Operands:** `inputs: Vec<Index>`. **Outputs:** 2 — a Jubjub point derived
  by hashing the inputs onto the curve.

### 3.8 I/O, transcript, and public inputs

#### `PublicInput { guard }`
* **Operands:** `guard: Option<Index>`. **Outputs:** 1.
* **Semantics:** witnesses the next value from the **public transcript
  outputs** into memory. If `guard` is present and `false`, pushes `0` and
  does **not** advance the transcript cursor.

#### `PrivateInput { guard }`
* **Operands:** `guard: Option<Index>`. **Outputs:** 1.
* **Semantics:** as `PublicInput`, but pulls from the **private transcript**
  (witness-only) stream.

#### `DeclarePubInput { var }`
* **Operands:** `var: Index`. **Outputs:** none.
* **Semantics:** appends `memory[var]` to the public-input vector and

>   advances the public-transcript-inputs cursor. Should be **followed** by a

Reviewer 2: Compact is assumed to include `PiSkip`s after public input declaration. It would be better not having to rely on these assumptions, but it is what it is.

```suggestion
  advances the public-transcript-inputs cursor. Must be **followed** by a
```
